HPCC-19970 In simplified expr use no_simplified operator

shamser opened the pull request:
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

shamser committed:
HPCC-19970 In simplified expr use no_simplified operator

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d3e24e9cc4b9844c00bc9ae1be4c9e24f31bf6d9
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |797 |797 |0 |
|test (thor) |728 |728 |0 |
|test (roxie) |868 |868 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   25 sec (00:00:25) |  168 sec (00:02:48) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1295 sec (00:21:35) |   17 sec (00:00:17) | 1602 sec (00:26:42) |

shamser commented:
@ghalliday Please can you review

ghalliday commented on ecl/hql/hqlcache.cpp:
```diff
@@ -355,29 +355,22 @@ static IHqlExpression * createSimplifiedDefinitionFromType(ITypeInfo * type, boo
         if (!expr->isFunctionDefinition())
             return nullptr;
-        OwnedHqlExpr newBody = createSimplifiedBodyDefinition(expr->queryChild(0)->queryBody());
+        LinkedHqlExpr newBody = createSimplifiedBodyDefinition(expr->queryChild(0)->queryBody());
```
This will leak memory

ghalliday commented on ecl/hql/hqlgram2.cpp:
```diff
@@ -12369,6 +12370,16 @@ bool verifySimpifiedDefinition(IHqlExpression *origExpr, IHqlExpression *simplif
     OwnedHqlExpr parsed = parseDefinition(ecl, simplifiedDefinition->queryId(), errors);
     if (!parsed)
     {
+#ifdef DEBUG
```
Should be have a leading underscore: _DEBUG

ghalliday commented on ecl/hql/hqlgram2.cpp:
```diff
@@ -12380,12 +12391,16 @@ bool verifySimpifiedDefinition(IHqlExpression *origExpr, IHqlExpression *simplif
 
         if (!streq(t1, t2))
         {
+#ifdef DEBUG
```
same here and following

ghalliday commented on ecl/hql/hqlir.cpp:
```diff
@@ -2031,6 +2032,7 @@ id_t ExpressionIRPlayer::doProcessExpr(IHqlExpression * expr)
         {
             HqlExprArray scopeSymbols;
             expr->queryScope()->getSymbols(scopeSymbols);
+            scopeSymbols.sort(compareSymbolsByName);
```
This has already been included in another PR - should really be rebasd on top of the current master

ghalliday reviewed (commented):
A few initial comments.  I'll run a regression test as well and see if I spot any other issues.

ghalliday commented:
@shamser

shamser commented:
@ghalliday I've fixed those issues mentioned in your comments. I added an additional change:  "Unable to expand constant expression when using cache..." is only shown if syntax checking is enabled AND caching is enabled

ghalliday commented:
@shamser a few more comments from running the regression suite:

codes2 shows a new entry for SIMPLFIED in an error message.  If should be included in the simplify list for expression hqlgram2.cpp(11408)

eclcc -internal reports an error:
    Error: 'simplified' is missing from reservedWords.cpp

digraph.ecl fails since SIMPLIFIED is no a reserved keyword.  I think you need to use __SIMPLIFIED__ instead to ensure that there are no other false clashes.

shamser commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e6a0b01b6fd30cc4ff89acd5b9a46c5221fe31f1
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  171 sec (00:02:51) |   62 sec (00:01:02) |    6 sec (00:00:06) |   27 sec (00:00:27) | 1326 sec (00:22:06) |   18 sec (00:00:18) | 1642 sec (00:27:22) |

ghalliday commented:
@shamser You also need the keyword from SIMPLIFIED to something that will not cause clashes with existing code e.g., \_\_SIMPLIFIED\_\_.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: de3c2c54f77e85226a89588a0785baf1c571c9a2
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  172 sec (00:02:52) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1301 sec (00:21:41) |   17 sec (00:00:17) | 1607 sec (00:26:47) |

shamser commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3c56884e6f0d1ffea1199018190647b427acc7c9
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   36 sec (00:00:36) |  171 sec (00:02:51) |   62 sec (00:01:02) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1304 sec (00:21:44) |   18 sec (00:00:18) | 1630 sec (00:27:10) |

ghalliday commented:
@shamser looks ready.  Please squash and I will merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 830c5b97c9d3f655469597f446b19bd8728cc7d2
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  173 sec (00:02:53) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1296 sec (00:21:36) |   19 sec (00:00:19) | 1601 sec (00:26:41) |
